libsql-sqlite3/src/vectorfloat8.c

```diff
@@ -0,0 +1,150 @@
+** Utility routines for vector serialization and deserialization
+**************************************************************************/
+
+void vectorF8GetParameters(const u8 *pData, int dims, float *pAlpha, float *pShift){
+  pData = pData + ALIGN(dims, sizeof(float));
+  *pAlpha = deserializeF32(pData);
+  *pShift = deserializeF32(pData + sizeof(*pAlpha));
+}
+
+void vectorF8SetParameters(u8 *pData, int dims, float alpha, float shift){
+  pData = pData + ALIGN(dims, sizeof(float));
+  serializeF32(pData, alpha);
+  serializeF32(pData + sizeof(alpha), shift);
+}
+
+void vectorF8Dump(const Vector *pVec){
+  u8 *elems = pVec->data;
+  float alpha, shift;
+  unsigned i;
+
+  assert( pVec->type == VECTOR_TYPE_FLOAT8 );
+
+  vectorF8GetParameters(pVec->data, pVec->dims, &alpha, &shift);
+
+  printf("f8: [");
+  for(i = 0; i < pVec->dims; i++){
+    printf("%s%f", i == 0 ? "" : ", ", (float)elems[i] * alpha + shift);
+  }
+  printf("]\n");
+}
+
+void vectorF8SerializeToBlob(
+  const Vector *pVector,
+  unsigned char *pBlob,
+  size_t nBlobSize
+){
+  float alpha, shift;
+
+  assert( pVector->type == VECTOR_TYPE_FLOAT8 );
+  assert( pVector->dims <= MAX_VECTOR_SZ );
+  assert( nBlobSize >= vectorDataSize(pVector->type, pVector->dims) );
+
+  memcpy(pBlob, pVector->data, pVector->dims);
+
+  vectorF8GetParameters(pVector->data, pVector->dims, &alpha, &shift);
+  vectorF8SetParameters(pBlob, pVector->dims, alpha, shift);
+}
+
+float vectorF8DistanceCos(const Vector *v1, const Vector *v2){
+  int i;
+  float alpha1, shift1, alpha2, shift2;
+  u32 sum1 = 0, sum2 = 0, sumsq1 = 0, sumsq2 = 0, doti = 0;
+  float dot = 0, norm1 = 0, norm2 = 0;
+  u8 *data1 = v1->data, *data2 = v2->data;
+
+  assert( v1->dims == v2->dims );
+  assert( v1->type == VECTOR_TYPE_FLOAT8 );
+  assert( v2->type == VECTOR_TYPE_FLOAT8 );
+
+  vectorF8GetParameters(v1->data, v1->dims, &alpha1, &shift1);
+  vectorF8GetParameters(v2->data, v1->dims, &alpha2, &shift2);
+
+  /*
+   * (Ax + S)^2 = A^2 x^2 + S^2 + 2AS x -> we need to maintain 'sumsq' and 'sum'
+   * (A1x + S1) * (A2y + S2) = A1A2 xy + A1 S2 x + A2 S1 y + S1 S2 -> we need to maintain 'dot' and 'sum' again
+  */
+
+  for(i = 0; i < v1->dims; i++){
+    sum1 += data1[i];
+    sum2 += data2[i];
+    sumsq1 += data1[i]*data1[i];
+    sumsq2 += data2[i]*data2[i];
+    doti += data1[i] * data2[i];
+  }
+
+  dot = alpha1 * alpha2 * (float)doti + alpha1 * shift2 * (float)sum1 + alpha2 * shift1 * (float)sum2 + shift1 * shift2;
+  norm1 = alpha1 * alpha1 * (float)sumsq1 + 2 * alpha1 * shift1 * (float)sum1 + shift1 * shift1;
+  norm2 = alpha2 * alpha2 * (float)sumsq2 + 2 * alpha2 * shift2 * (float)sum2 + shift2 * shift2;
+
+  return 1.0 - (dot / sqrt(norm1 * norm2));
+}
+
+float vectorF8DistanceL2(const Vector *v1, const Vector *v2){
+  assert( v1->dims == v2->dims );
+  assert( v1->type == VECTOR_TYPE_FLOAT8 );
+  assert( v2->type == VECTOR_TYPE_FLOAT8 );
+
+  assert( 0 );
```

Comment:
Why does this always fail?

Reply:
Do you look at the latest commit from the branch...?

Reply:
I was checking first commit. Nvm